### DISCUSSION TITLE
Tracing integration with unify.ai

### DISCUSSION THREAD

**Discussion by marcklingen** [langfuse/langfuse#2241] [Ideas]
### Describe the feature or potential improvement

https://unify.ai is a commercial prompt/model router

**Comment by KatoStevenMubiru** [comment]
Hi Marc and team,

Thank you for initiating this discussion. We have two potential approaches for integrating Unify.ai with Langfuse and would appreciate your feedback.

Implementation 1:
Link: https://github.com/KatoStevenMubiru/langfuse-python/blob/main/langfuse/unify.py
Description: Integrates Unify.ai by defining methods for synchronous and asynchronous operations, initializing Langfuse with Unify configurations, and wrapping Unify methods for usage tracking and cost logging.

Pros: Optimized performance, easy integration.
Cons: Requires additional configuration, limited edge case support.

Implementation 2:
Link: https://github.com/Anteemony/langfuse-python/blob/unify-integration/langfuse/unify/unify_integration.py
Description: Replaces unify import with langfuse.unify for full logging and tracking (prompts, completions, latencies, errors, usage, cost). Uses observe() decorator and tracing SDK.

Pros: Simpler, more flexible.
Cons: Potentially higher latency, more initial setup.

Request for Feedback:
Which implementation aligns better with Langfuse’s standards?
Any recommended improvements or adjustments?
Additional features to enhance the integration?
We are open to further discussions if needed.

Best regards,

Kato Steven Mubiru

**Comment by Kacper-W-Kozdon** [comment]
Current integration suggestions (which still need some work) are Pyton-based. We received a suggestion to also look into integrating with the TypeScript part of LangFuse- but right now I'm not sure how to approach that. As a sidenote, Unify can be used with the openAI objects just by replacing the url in openai to that of Unify.

**Reply by Kacper-W-Kozdon** [reply]
Hi @marcklingen , I think [this integration proposal](https://github.com/KatoStevenMubiru/langfuse-python/tree/main/langfuse/unify) can be slowly looked into to check if the direction is right. So far, it's borrowing a lot from openai's implementation (Unify's backbone is openai). Some things need to be ironed out still: like the format of the arguments passed to LangFuse. I also want to check in with Dan to figure out, if Unify's dynamic routing can be captured by langfuse (through `langfuse.event`? openai doesn't seem to use anything beyond `trace` and `generation`). But this is the more complete approach, tracing unify and with the possibility to edit which elements from unify get traced.

There is also a simpler solution: taking openai integration and just replacing the url to that of unify. It will still do unify's things but with openai syntax and if unify receives new features, that won't be tracked. Or loading `langfuse.openai` as `openai` into unify- that would also do the job but again- without tracking any future changes to unify.

**Comment by hassiebp** [comment]
Hi @Kacper-W-Kozdon and @KatoStevenMubiru - thanks a lot for this intiative to integrate Langfuse with Unify! 

- I took a brief look at your implementation, and it looks directionally right
- However, at the current time we do not plan to merge this integration into the main Langfuse project. We would prefer that such integrations be maintained in your own package. This approach allows for more flexibility and autonomy for you to manage and update the integration as needed. 
- For reference, please check out the [Haystack integration](https://github.com/deepset-ai/haystack-core-integrations/tree/main/integrations/langfuse) with Langfuse that are maintained in their own packages

Good luck on building this out and feel free to send a PR into your own packages our way for us to review once this takes a final shape.

**Reply by WHITELOTUS0** [reply]
Alright, thanks for the response
